### DISCUSSION TITLE
How-to guide on updating database periodically

### DISCUSSION THREAD

**Discussion by MarcSkovMadsen** [duckdb/duckdb#12855] [Ideas]
I'm building a data app using in Python on top of duck db.

Its unclear to me what the best practice for updating the database periodically is. Please document how to update an in use database periodically on a schedule for Python users.

## My use case

The source files for my database a parquet files in a blob storage. For simplicity lets assume I have a single, small jobs.parquet file keeping track of jobs by `jobid` and a very large `timeseries.parquet` file that is hive partitioned by `job_name` and `notationtime`.

On a schedule an airflow job updates both parquet files.

My first, preliminary update attempt is just to replace the local `dashboard.db` file in its entirety by an external airflow job. But its very unclear to me what will happen for existing duckdb connetions? They seem continue to work and use data from the original `.db` file. But if and when this will break I don't know.

Its unclear for me if I should continue to go down the replace db in entirety route maybe making sure connections are reopened if there is a not `.db` file. Or if I should try to upsert data into the existing database on a schedule. And whether I need to do that from a process inside the application or whether I can do that from an outside (airflow) job.

I think this use case will be the same for data apps in [Panel](https://panel.holoviz.org/index.html), Streamlit, Dash, Gradio etc and for REST APIs too? I use Panel for my data app.

## My questions

- May I replace the `.db` file in its entirety?
  - Should I close/ delete existing connections? Or can they continue to work?
- Is best practice to replace `.db` file or update existing `.db` file?`
- Can I update the `.db` file from an external (airflow) job while an application is using the `.db` file?
- Where do I find a minimal, reproducible example of this pattern?

## Contribution

If someone outlines how this should be done. I might be interested in contributing a how-to guide.

**Comment by Tishj** [comment]
I'd be happy to answer the outlined questions, but first I have a clarifying question.
> The source files for my database a parquet files in a blob storage

How does this translate to a duckdb file?
It sounds like the database file would only contain VIEWs, so the database wouldn't need to be recreated if only the content of the referenced parquet files changed

**Comment by Tishj** [comment]
I don't think replacing the `.db` file is recommended.
I'm not quite sure on the details but I imagine the file doesn't get unlinked by the filesystem until the process using it has closed it.

I think all of this is covered by <https://duckdb.org/docs/connect/concurrency#writing-to-duckdb-from-multiple-processes>
But to reiterate, you can open a db for reading+writing with a single process, or you can open a db read-only with multiple processes.

If you're updating the db file, that constitutes a write operation and there can be no other processes using the db during that time, not even if those are read-only connections.
